druid/src/text/rich_text.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright 2018 The Druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Rich text with style spans.
+
+use std::ops::{Range, RangeBounds};
+use std::sync::Arc;
+
+use super::{Attribute, AttributeSpans, TextStorage};
+use crate::piet::{
+    util, Color, FontFamily, FontStyle, FontWeight, PietTextLayoutBuilder, TextLayoutBuilder,
+    TextStorage as PietTextStorage,
+};
+use crate::{ArcStr, Data, Env, FontDescriptor, KeyOrValue};
+
+/// Text with optional style spans.
+#[derive(Debug, Clone, Data)]
+pub struct RichText {
+    buffer: ArcStr,
+    attrs: Arc<AttributeSpans>,
+}
+
+impl RichText {
+    /// Create a new `RichText` object with the provided text.
+    pub fn new(buffer: ArcStr) -> Self {
+        RichText::new_with_attributes(buffer, Default::default())
+    }
+
+    /// Create a new `RichText`, providing explicit attributes.
+    pub fn new_with_attributes(buffer: ArcStr, attributes: AttributeSpans) -> Self {
+        RichText {
+            buffer,
+            attrs: Arc::new(attributes),
+        }
+    }
+
+    /// Builder-style method for adding an [`Attribute`] to a range of text.
+    ///
+    /// [`Attribute`]: enum.Attribute.html
+    pub fn with_attribute(mut self, range: impl RangeBounds<usize>, attr: Attribute) -> Self {
+        self.add_attribute(range, attr);
+        self
+    }
+
+    /// The length of the buffer, in utf8 code units.
+    pub fn len(&self) -> usize {
+        self.buffer.len()
+    }
+
+    /// Returns `true` if the underlying buffer is empty.
+    pub fn is_empty(&self) -> bool {
+        self.buffer.is_empty()
+    }
+
+    /// Add an [`Attribute`] to the provided range of text.
+    ///
+    /// [`Attribute`]: enum.Attribute.html
+    pub fn add_attribute(&mut self, range: impl RangeBounds<usize>, attr: Attribute) {
+        let range = util::resolve_range(range, self.buffer.len());
+        Arc::make_mut(&mut self.attrs).add(range, attr);
+    }
+}
+
+impl PietTextStorage for RichText {
+    fn as_str(&self) -> &str {
+        self.buffer.as_str()
+    }
+}
+
+impl TextStorage for RichText {
+    fn add_attributes(
+        &self,
+        mut builder: PietTextLayoutBuilder,
+        env: &Env,
+    ) -> PietTextLayoutBuilder {
+        for (range, attr) in self.attrs.to_piet_attrs(env) {
+            builder = builder.range_attribute(range, attr);
+        }
+        builder
+    }
+}
+
+/// A builder for [`RichText`].
+///
```

Comment:
I'd flesh this out a little bit more, I found it slightly confusing:

```suggestion
/// A builder for creating [`RichText`] objects.
///
/// This builder allows you to construct a [`RichText`] object by building up a sequence
/// of styled sub-strings; first you [`push`](RichText::push) a `&str` onto the string,
/// and then you can optionally add styles to that text via the returned [`AttributesAddr`]
/// object.
```